src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractGitService.java

```diff
@@ -0,0 +1,337 @@
+import org.eclipse.jgit.transport.CredentialsProvider;
+import org.eclipse.jgit.transport.SshConfigStore;
+import org.eclipse.jgit.transport.SshConstants;
+import org.eclipse.jgit.transport.SshTransport;
+import org.eclipse.jgit.transport.URIish;
+import org.eclipse.jgit.transport.sshd.KeyPasswordProvider;
+import org.eclipse.jgit.transport.sshd.SshdSessionFactory;
+import org.eclipse.jgit.transport.sshd.SshdSessionFactoryBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+
+import de.tum.in.www1.artemis.domain.Repository;
+import de.tum.in.www1.artemis.domain.VcsRepositoryUri;
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
+
+public abstract class AbstractGitService {
+
+    private static final Logger log = LoggerFactory.getLogger(AbstractGitService.class);
+
+    protected TransportConfigCallback sshCallback;
+
+    protected static final int JGIT_TIMEOUT_IN_SECONDS = 5;
+
+    protected static final String REMOTE_NAME = "origin";
+
+    @Value("${artemis.version-control.url}")
+    protected URL gitUrl;
+
+    @Value("${artemis.version-control.user}")
+    protected String gitUser;
+
+    @Value("${artemis.version-control.password}")
+    protected String gitPassword;
+
+    @Value("${artemis.version-control.token:#{null}}")
+    protected Optional<String> gitToken;
+
+    @Value("${artemis.version-control.ssh-private-key-folder-path:#{null}}")
+    protected Optional<String> gitSshPrivateKeyPath;
+
+    @Value("${artemis.version-control.ssh-private-key-password:#{null}}")
+    protected Optional<String> gitSshPrivateKeyPassphrase;
+
+    @Value("${artemis.version-control.ssh-template-clone-url:#{null}}")
+    protected Optional<String> sshUrlTemplate;
+
+    @Value("${artemis.version-control.default-branch:main}")
+    protected String defaultBranch;
+
+    protected AbstractGitService() {
+        log.debug("file.encoding={}", Charset.defaultCharset().displayName());
+        log.debug("sun.jnu.encoding={}", System.getProperty("sun.jnu.encoding"));
+        log.debug("Default Charset={}", Charset.defaultCharset());
+        log.debug("Default Charset in Use={}", new OutputStreamWriter(new ByteArrayOutputStream()).getEncoding());
+    }
+
+    protected static TransportConfigCallback getSshCallback(SshdSessionFactory sshSessionFactory) {
+        return transport -> {
+            if (transport instanceof SshTransport sshTransport) {
+                transport.setTimeout(JGIT_TIMEOUT_IN_SECONDS);
+                sshTransport.setSshSessionFactory(sshSessionFactory);
+            }
+            else {
+                log.error("Cannot use ssh properly because of mismatch of Jgit transport object: {}", transport);
+            }
+        };
+    }
+
+    protected static SshdSessionFactoryBuilder getSshdSessionFactoryBuilder(Optional<String> gitSshPrivateKeyPath, Optional<String> gitSshPrivateKeyPassphrase, URL gitUrl) {
```

Comment:
Can you extract the inner classes from this method? It's quite hard to follow which method belongs to what